src/dsl/static.jl

```diff
@@ -66,184 +66,172 @@ function parse_julia_expr!(stmts, bindings, name::Symbol, var::Symbol,
+    node = gen_node_name(addr) # Generate a variable name for the StaticIRNode
+    bindings[name] = node
+    if !isa(fn, Symbol)
+        static_dsl_syntax_error(expr_s, "$fn is not a Symbol")
     end
-    @assert isa(addr_expr.args[2], LineNumberNode)
-    call = addr_expr.args[3]
-    if !isa(call, Expr) || call.head != :call
-        return false
+    gen_fn_or_dist = gensym(fn)
+    push!(stmts, :($(esc(gen_fn_or_dist)) = $(esc(fn))))
+
+    keys = []
+    split_addr!(keys, addr) # Split nested addresses
+    if !(isa(keys[1], QuoteNode) && isa(keys[1].value, Symbol))
+        static_dsl_syntax_error(addr, "$(keys[1].value) is not a Symbol")
     end
-    local addr::Symbol
-    gen_fn_or_dist = gensym()
-    push!(stmts, :($(esc(gen_fn_or_dist)) = $(esc(call.args[1]))))
-    if isa(addr_expr.args[4], QuoteNode) && isa(addr_expr.args[4].value, Symbol)
-        addr = addr_expr.args[4].value
-        args = call.args[2:end]
-    else
-        # multi-part address syntactic sugar
-        keys = []
-        split_addr!(keys, addr_expr.args[4])
-        if !isa(keys[1], QuoteNode) || !isa(keys[1].value, Symbol)
-            return false
-        end
-        @assert length(keys) > 1
-        addr = keys[1].value
+    addr = keys[1].value # Get top level address
+    if length(keys) > 1
         for key in keys[2:end]
-            push!(stmts, :($(esc(gen_fn_or_dist)) = choice_or_call_at($(esc(gen_fn_or_dist)), Any)))
+            # For each nesting level, wrap fn within choice_at / call_at
+            push!(stmts, :($(esc(gen_fn_or_dist)) =
+                choice_or_call_at($(esc(gen_fn_or_dist)), Any)))
         end
-        args = (call.args[2:end]..., reverse(keys[2:end])...)
+        # Append the nested addresses as arguments to choice_at / call_at
+        args = [args; reverse(keys[2:end])]
     end
-    node = gensym()
-    if haskey(bindings, name)
-        static_dsl_syntax_error(addr_expr, "Symbol $name already bound")
-    end
-    bindings[name] = node
+
     inputs = []
     for arg_expr in args
-        push!(inputs, parse_julia_expr!(stmts, bindings, gensym(), arg_expr, QuoteNode(Any)))
+        if MacroTools.@capture(arg_expr, x_...)
+            static_dsl_syntax_error(expr_s, "Cannot splat in @trace call.")
+        end
+        # Create Julia node for each argument to gen_fn_or_dist
+        arg_name = gen_node_name(arg_expr)
+        push!(inputs, parse_julia_expr!(stmts, bindings,
+                                        arg_name, arg_expr, QuoteNode(Any)))
     end
+
+    # Add addr node
     push!(stmts, :($(esc(node)) = add_addr_node!(
-        builder, $(esc(gen_fn_or_dist)), inputs=[$(map(esc, inputs)...)], addr=$(QuoteNode(addr)),
-        name=$(QuoteNode(name)))))
-    true
+        builder, $(esc(gen_fn_or_dist)), inputs=[$(map(esc, inputs)...)],
+        addr=$(QuoteNode(addr)), name=$(QuoteNode(name)))))
+    # Return the name of the newly created node
+    return name
```

Comment:
@ztangent And this is the code that, combined with the bottom-up traversal and replacement, allows you to nest `@trace` expressions inside Julia expressions?

Reply:
Yup, exactly, `return name` does that.